bootstrap-languages/language-language/hc-dna/zomes/language_storage/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+use chrono::{DateTime, NaiveDateTime, Utc};
+use hdk::prelude::*;
+use integrity::{EntryTypes, LanguageChunk, LanguageExpression, LinkTypes, LanguageAddress};
+mod utils;
+use utils::{err, get_latest_link};
+
+#[hdk_extern]
+fn init(_: ()) -> ExternResult<InitCallbackResult> {
+    Ok(InitCallbackResult::Pass)
+}
+
+pub fn get_now() -> DateTime<Utc> {
+    match sys_time() {
+        Ok(time) => {
+            let now = time.as_seconds_and_nanos();
+            let out = DateTime::<Utc>::from_utc(
+                NaiveDateTime::from_timestamp_opt(now.0, now.1).unwrap(),
+                Utc,
+            );
+            out
+        }
+        Err(_err) => Utc::now(),
+    }
+}
+
+#[hdk_extern]
+pub fn store_language_expression(expression: LanguageExpression) -> ExternResult<()> {
+    let address = EntryTypes::LanguageAddress(LanguageAddress(expression.data.address.clone()));
+    let address_hash = hash_entry(&address)?;
+
+    let language_expression = EntryTypes::LanguageExpression(expression);
+    let language_expression_hash = hash_entry(&language_expression)?;
+    create_entry(&language_expression)?;
+
+    //Link profile entry to did
+    create_link(
+        address_hash,
+        language_expression_hash,
+        LinkTypes::LanguageLink,
+        LinkTag::from("".as_bytes().to_owned()),
+    )?;
+
+    Ok(())
+}
+
+#[hdk_extern]
+pub fn store_chunk(file_chunk: LanguageChunk) -> ExternResult<EntryHash> {
+    let file_chunk_hash = hash_entry(&file_chunk)?;
+
+    if let None = get(file_chunk_hash.clone(), GetOptions::default())? {
+        create_entry(&EntryTypes::LanguageChunk(file_chunk))?;
+    }
+
+    Ok(file_chunk_hash)
+}
+
+#[hdk_extern]
+pub fn get_language_expression(
+    file_expression_hash: LanguageAddress,
+) -> ExternResult<Option<LanguageExpression>> {
+    let expression_links = get_latest_link(
```

Comment:
With this and without validation rules it is possible to publish the same language (same code file) multiple times from multiple agents which causes problems with trusted agents (we had that problem some months ago with the centralised language language).

An easy fix could be to not take the latest link but that oldest link here.

A better solution would be to make it impossible to add a second link off of a LanguageAddress anchor.